Add System.Drawing.Icon tests

hughbe opened the pull request:
Contributes to #20711
Fixes #20808

This is a 50-50 mixture of handwritten tests as well as tests inspired and cleanup up from Mono

danmosemsft commented on src/System.Drawing.Common/tests/IconTests.cs:
```diff
@@ -0,0 +1,586 @@
+﻿using System.Collections.Generic;
```
@hughbe as you probably know if we copy code from Mono we have to mess with the copyright header as below (also we need one :))

```
	1. Keep the existing copyright headers in place
		○ Note: If you are porting just portions of the file over, you still need to bring over all the existing copyright headers.
	2. Add the following copyright headers – NOTE this is missing the middle line of the ‘regular’ CoreFx copyright headers:
		// Licensed to the .NET Foundation under one or more agreements.
                // See the LICENSE file in the project root for more information.
```

If it's bits of Mono or inspired by bits of Mono - well I'm not a lawyer but it seems to me we should do the same thing.

mellinoe commented on src/System.Drawing.Common/tests/Helpers.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        public static string GetEmbeddedResourcePath(string name)
+        {
+            return Path.Combine(Assembly.GetExecutingAssembly().Location, "..", "bitmaps", name);
```
Is there a significance to these being embedded resources if they are just copied to the output directory and used as files?

mellinoe replied:
Small nit: I would prefer this to be

`Path.Combine(AppContext.BaseDirectory, "bitmaps", name)`

Effectively it's the same, but doesn't have the unusual ".." in the path.

hughbe replied:
Sure thing, I did this

hughbe replied:
Sure thing, I did this

mellinoe commented:
We will need to exclude these tests on Nano; the implementation doesn't work there.

norek commented:
Oh I see that you also add test project. Ill wait with my commits for your merge to prevent possible merge conflicts

mellinoe commented:
@norek  Thanks for the patience. Once we get the test project in place we should be able to work in parallel without having to worry about merge conflicts.

mellinoe commented on src/System.Drawing.Common/tests/System.Drawing.Common.Tests.csproj:
```diff
@@ -0,0 +1,57 @@
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+      <LogicalName>System.Drawing.Tests.Smiley.ico</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="bitmaps\323511.ico">
```
Where did the names for these come from? It might be better if they all had descriptive names. "rounded_green_square_256x256" would work for this one, etc.

hughbe replied:
renamed all of the tests

mellinoe commented on src/System.Drawing.Common/tests/IconTests.cs:
```diff
@@ -0,0 +1,586 @@
+        [Fact]
+        public void Ctor_Stream()
+        {
+            using (var stream = new FileStream(Helpers.GetEmbeddedResourcePath("smiley.ico"), FileMode.Open))
```
This could be using `Assembly.GetManifestResourceStream`, though perhaps these files should just not be embedded resources in the first place.

hughbe replied:
Yeah I converted all of them to not be embedded apart from one, which is used to test the Ctor that loads embedded resources

mellinoe commented:
The tests themselves look good and comprehensive to me. Thanks for adding these, @hughbe !

danmosemsft commented:
@mellinoe @hughbe looks like at least parts of System.Drawing are not going to work right on Nano. I'm not sure how much of user32.dll exists there.

@safern is there any way to skip all System.Drawing tests for Nano without doing it on every single test method? We usually use ConditionalTheory but I do'nt know a way to conditialize a whole fixture.
```
System.EntryPointNotFoundException : Unable to find an entry point named 'CreateIconFromResourceEx' in DLL 'user32.dll'.
Stack Trace :
   at System.Drawing.SafeNativeMethods.IntCreateIconFromResourceEx(Byte* pbIconBits, Int32 cbIconBits, Boolean fIcon, Int32 dwVersion, Int32 csDesired, Int32 cyDesired, Int32 flags)
   at System.Drawing.Icon.Initialize(Int32 width, Int32 height)
   at System.Drawing.Tests.IconTests.Clone_ConstructedIcon_Success()
```

In other news, it would be great if one of you could open an issue to make S.Drawing itself throw a nice PNSE with a message on Nano. Maybe a convenient place is inside SafeNativeMethods functions.

mellinoe commented:
@danmosemsft Filed https://github.com/dotnet/corefx/issues/20834. We should wrap the native calls in some way so that the exception is better than just an EntryPointNotFoundException.

safern commented:
> @safern is there any way to skip all System.Drawing tests for Nano without doing it on every single test method? We usually use ConditionalTheory but I do'nt know a way to conditialize a whole fixture.

I'm not sure. Maybe through OSGroup or OSEnvironment in the csproj? @weshaggard do you know?

mellinoe commented:
I don't think there's a way to do it through the configuration system; there's just one Windows (non-UAP) build, and Nano is not a distinct platform there. I don't know if we've ever used any API's which are not available only on Nano.

danmosemsft commented:
@mellinoe we have at least one, ShellExecuteEx. In that case it happens there's a stub that returns fail but this is the typical case, EPNFE. Either way it needs special handling to give a nice PNSE.

hughbe commented:
I've addressed all the PR feedback, disabling Windows Nano tests and the like. I mentioned in another issue that I was going to hold on this PR, but I don't want to block myself if I want to add some more tests this weekend.

safern commented:
> /cc @safern no pressure

Thanks, I would love to do it tomorrow but I have some other higher priority stuff. I'll let you know when that is ready.

jamesqo commented on src/System.Drawing.Common/tests/IconTests.cs:
```diff
@@ -0,0 +1,614 @@
+        [InlineData("256x256_seven_entries_multiple_bits.ico")]
+        public void Ctor_FilePath(string name)
+        {
+            var icon = new Icon(Helpers.GetEmbeddedResourcePath("48x48_multiple_entries_4bit.ico"));
```
Should `name` be used here instead?

hughbe replied:
nice spot

mellinoe commented on src/System.Drawing.Common/tests/Helpers.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public static class Helpers
+    {
+        public static string GetEmbeddedResourcePath(string name)
```
The name of this function is a bit misleading, I think. Perhaps "GetTestBitmapPath()" would be better.

hughbe replied:
Sure, done

mellinoe commented:
I just pointed out a couple of nit-picky things. Beyond that, it looks good to me.

mellinoe commented on src/System.Drawing.Common/tests/IconTests.cs:
```diff
@@ -0,0 +1,614 @@
+        [InlineData(typeof(Icon), "")]
+        [InlineData(typeof(Icon), "48x48_multiple_entries_4bit.ico")]
+        [InlineData(typeof(IconTests), "48x48_MULTIPLE_entries_4bit.ico")]
+        public void Ctor_InvalidResource_ThrowsArgumentNullException(Type type, string resource)
```
The name of the method should be `Ctor_InvalidResource_ThrowsArgumentException` (not ArgumentNullException).

hughbe replied:
Fixed this and others

hughbe commented:
@mellinoe what order are you going to merge the PRs? I don't think this will conflict with the unix stuff but it will cause *some* test failures.

Also this will conflict with my pens/brushes PR, as they both change the csproj

mellinoe commented:
@dotnet-bot Test Portable Linux x64 Debug Build
Test Portable Linux x64 Release Build

mellinoe commented:
@hughbe The last few commits look good to me. I'm going to merge this once the last two legs finish re-running. I wasn't able to identify what went wrong in the runs previously.
